Accept the `created_via` param to the `POST /orders` endpoint

albarin opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR allows setting the `created_via` parameters for the `POST /wp-json/wc/v3/orders` endpoint.
When not sent, `created_via` will be set to the default `rest-api`.

Closes #56670

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to `wp-admin/admin.php?page=wc-settings&tab=advanced&section=keys` and add a new key with `Read/Write` permissions.
2. Make the following request using the consumer key and secret generated in the previous step, adding the `created_via` param in the body:
```bash
curl --insecure --request POST \
  --url https://[STORE_URL]/wp-json/wc/v3/orders \
  -u [KEY]:[SECRET] \
  --header 'Content-Type: application/json' \
  --data '{
  "created_via": "[somevalue]"
}'
```
3. Confirm the created order has the value you chose for the `created_via` parameter.
4. Run the same request again, but without the `created_via`.
5. Confirm the created order has the value `rest-api` in the `created_via` parameter.
6. Run a PUT request to update one of the orders you just created and try to update the `created_via` field:
```bash
curl --insecure --request PUT \
  --url https://[STORE_URL]/wp-json/wc/v3/orders/[ORDER-ID] \
  -u [KEY]:[SECRET] \
  --header 'Content-Type: application/json' \
  --data '{
  "created_via": "[a new value]"
}'
```
7. Make sure the `created_via` field is **NOT updated** and it still has the original value.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Accept the "created_via" param to the "POST /orders" endpoint.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

albarin committed:
Accept the `created_vid` param to the `POST /orders` endpoint

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

albarin committed:
Merge branch 'trunk' into 56670-accept-created_via-in-post-orders-endpoint

albarin committed:
Remove read-only property from created_via for `POST /orders`

malinajirka commented:
Thanks @albarin! I've tested the behavior, including requests through the jetpack tunnel, and everything seems to be working as expected.

I also tested some malicious entries and I'm wondering if the current behavior is expected or we need to do some extra sanitization. Both SQL injection and XSS kept the entries exactly as they were - I mean, I couldn't find an actual issue, since the code wasn't interpreted, but I'm wondering if it'd be safer to just escape these immediately.

albarin committed:
Add sanitization to created_via

albarin committed:
Prevent modifying created_via on update

albarin commented:
> I also tested some malicious entries and I'm wondering if the current behavior is expected or we need to do some extra sanitization. Both SQL injection and XSS kept the entries exactly as they were - I mean, I couldn't find an actual issue, since the code wasn't interpreted, but I'm wondering if it'd be safer to just escape these immediately.

@malinajirka Thanks, I think you are right, I've added some extra sanitization.

Also, I realised something was missing: we agreed to not allow updating `created_via` on `PUT /orders`, so I've pushed that change here https://github.com/woocommerce/woocommerce/pull/57225/commits/08053dc41f068eba428a6250624e55270e659184 and updated the testing steps. Could you have another look?

malinajirka reviewed (approved):
I've tested the changes and it seems to work as expected, thank you! It should still be reviewed by someone with more PHP experience.

kalessil commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-orders-controller.php:
```diff
@@ -118,6 +118,12 @@ protected function prepare_object_for_database( $request, $creating = false ) {
 
 			if ( ! is_null( $value ) ) {
 				switch ( $key ) {
+					case 'created_via':
```
I am puzzled a bit by the removal of `'readonly' => true` in controllers V1 and V2 without replicating the changes from this file in those controllers. What's our thinking behind this approach?

albarin replied:
Oops, you are right, that didn't make sense, I've reverted the change on https://github.com/woocommerce/woocommerce/pull/57225/commits/084ddf6cc8675a84ac40e0f615e8d99c40ba7d09

kalessil replied:
Looks good, thanks! We are good to merge.

kalessil reviewed (commented):
LGTM! Manual and unit tests are passing :heavy_check_mark: 

Leaving a question to build more context on how we updating core REST APIs.

albarin committed:
Make created_via readonly only for v3

albarin committed:
Merge branch 'trunk' into 56670-accept-created_via-in-post-orders-endpoint
